File: services/k8s-service-go/internal/routes/cluster.go
package routes

import (
	"github.com/go-chi/chi/v5"

	"github.com/junginho0901/kubeast/services/k8s-service-go/internal/handler"
)

// RegisterCluster — cluster-scoped overview, namespaces, nodes, events,
// resource graph / dependency graph, timeline, topology.
func RegisterCluster(r chi.Router, h *handler.Handler) {
	// Overview & cluster
	r.Get("/api/v1/overview", h.GetOverview)
	r.Get("/api/v1/api-resources", h.GetAPIResources)
	r.Get("/api/v1/cluster-config", h.GetClusterConfig)
	r.Get("/api/v1/componentstatuses", h.GetComponentStatuses)

	// Namespaces
	r.Get("/api/v1/namespaces", h.GetNamespaces)
	r.Post("/api/v1/namespaces", h.CreateNamespace)
	r.Get("/api/v1/namespaces/{namespace}/describe", h.DescribeNamespace)
	r.Get("/api/v1/namespaces/{namespace}/yaml", h.GetNamespaceYAML)
	r.Post("/api/v1/namespaces/{namespace}/yaml/apply", h.ApplyNamespaceYAML)
	r.Delete("/api/v1/namespaces/{namespace}", h.DeleteNamespace)
	r.Get("/api/v1/namespaces/{namespace}/resource-quotas", h.GetNamespaceResourceQuotas)
	r.Get("/api/v1/namespaces/{namespace}/limit-ranges", h.GetNamespaceLimitRanges)
	r.Get("/api/v1/namespaces/{namespace}/owned-pods", h.GetNamespaceOwnedPods)

	// Nodes
	r.Get("/api/v1/nodes", h.GetNodes)
	r.Get("/api/v1/nodes/{name}/describe", h.DescribeNode)
	r.Get("/api/v1/nodes/{name}/yaml", h.GetNodeYAML)
	r.Get("/api/v1/nodes/{name}/pods", h.GetNodePods)
	r.Get("/api/v1/nodes/{name}/events", h.GetNodeEvents)
	r.Delete("/api/v1/nodes/{name}", h.DeleteNode)
	r.Post("/api/v1/nodes/{name}/yaml/apply", h.ApplyNodeYAML)
	r.Post("/api/v1/nodes/{name}/cordon", h.CordonNode)
	r.Post("/api/v1/nodes/{name}/uncordon", h.UncordonNode)
	r.Post("/api/v1/nodes/{name}/drain", h.DrainNode)
	r.Get("/api/v1/nodes/{name}/drain/status", h.DrainNodeStatus)
	r.Get("/api/v1/nodes/{name}/debug-shell/ws", h.NodeDebugShellWS)

	// Events
	r.Get("/api/v1/events", h.GetEvents)
	r.Get("/api/v1/namespaces/{namespace}/events", h.GetNamespaceEvents)

	// Dependency Graph (legacy)
	r.Get("/api/v1/namespaces/{namespace}/dependency-graph", h.GetDependencyGraph)

	// Resource Graph (upgraded)
	r.Get("/api/v1/resource-graph", h.GetResourceGraph)
	r.Get("/api/v1/namespaces/{namespace}/resource-graph", h.GetNamespaceResourceGraph)

	// Timeline (nginx rewrites /api/v1/cluster/namespaces/* → /api/v1/namespaces/*)
	r.Get("/api/v1/namespaces/{namespace}/timeline", h.GetNamespaceTimeline)
	r.Get("/api/v1/namespaces/{namespace}/timeline/{kind}/{name}", h.GetResourceTimeline)

	// Topology
	r.Get("/api/v1/topology/namespace/{namespace}", h.GetNamespaceTopology)
	r.Get("/api/v1/topology/service/{namespace}/{service_name}", h.GetServiceTopology)
	r.Get("/api/v1/topology/deployment/{namespace}/{deployment_name}", h.GetDeploymentTopology)
	r.Get("/api/v1/topology/storage", h.GetStorageTopology)
}

File: frontend/src/services/api/cluster.ts
// Cluster API — overview / namespaces / nodes / api-resources /
// component statuses / generic resource (yaml/json) / search.
// Pod / workload / network / storage etc. live in their own
// domain files even though the gateway URL is rooted at /cluster.

import axios from 'axios'

import { client } from './client'
import type {
  ClusterOverview,
  NamespaceDescribe,
  NamespaceInfo,
  NamespaceLimitRange,
  NamespacePod,
  NamespaceResourceQuota,
  PodInfo,
} from './types'

export const clusterApi = {
  getClusterOverview: async (forceRefresh = false): Promise<ClusterOverview> => {
    const { data } = await client.get('/cluster/overview', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getNamespaces: async (forceRefresh = false): Promise<NamespaceInfo[]> => {
    const { data } = await client.get('/cluster/namespaces', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeNamespace: async (namespace: string): Promise<NamespaceDescribe> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/describe`)
    return data
  },

  getNamespaceYaml: async (name: string, forceRefresh: boolean = false): Promise<{ yaml: string }> => {
    const { data } = await client.get(`/cluster/namespaces/${name}/yaml`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  applyNamespaceYaml: async (name: string, yaml: string): Promise<{ status: string }> => {
    const { data } = await client.post(`/cluster/namespaces/${name}/yaml/apply`, { yaml })
    return data
  },

  getNamespaceResourceQuotas: async (namespace: string): Promise<NamespaceResourceQuota[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/resource-quotas`)
    return data
  },

  getNamespaceLimitRanges: async (namespace: string): Promise<NamespaceLimitRange[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/limit-ranges`)
    return data
  },

  getNamespacePods: async (namespace: string): Promise<NamespacePod[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/owned-pods`)
    return data
  },

  createNamespace: async (name: string): Promise<{ status: string; name: string }> => {
    const { data } = await client.post('/cluster/namespaces', { name })
    return data
  },

  deleteNamespace: async (name: string): Promise<{ status: string; name: string }> => {
    const { data } = await client.delete(`/cluster/namespaces/${name}`)
    return data
  },

  // Generic resource (yaml / json / apply / multi-create) — shared
  // pathway used by ResourceDetailDrawer / yaml editors. Stays here
  // because the URL surface is /cluster/resources/*.
  getResourceYaml: async (resourceType: string, name: string, namespace?: string): Promise<{ yaml: string }> => {
    const { data } = await client.get('/cluster/resources/yaml', {
      params: {
        resource_type: resourceType,
        resource_name: name,
        ...(namespace && namespace !== '-' ? { namespace } : {}),
      },
    })
    return typeof data === 'string' ? { yaml: data } : data
  },

  getResourceJson: async (resourceType: string, name: string, namespace?: string): Promise<Record<string, unknown>> => {
    const { data } = await client.get('/cluster/resources/json', {
      params: {
        resource_type: resourceType,
        resource_name: name,
        ...(namespace && namespace !== '-' ? { namespace } : {}),
      },
    })
    return data
  },

  applyResourceYaml: async (resourceType: string, name: string, yaml: string, namespace?: string): Promise<{ status: string }> => {
    const { data } = await client.post('/cluster/resources/yaml/apply', {
      resource_type: resourceType,
      resource_name: name,
      namespace: namespace && namespace !== '-' ? namespace : undefined,
      yaml,
    })
    return data
  },

  createResourcesFromYaml: async (
    yaml: string,
    namespace?: string,
  ): Promise<{
    status: string
    count: number
    created: Array<{ apiVersion: string; kind: string; name: string; namespace?: string | null }>
  }> => {
    const { data } = await client.post('/cluster/resources/yaml/create', {
      yaml,
      namespace: namespace && namespace !== '-' ? namespace : undefined,
    })
    return data
  },

  // Cluster View — node + cluster-wide
  getNodes: async (forceRefresh: boolean = false): Promise<any[]> => {
    const { data } = await client.get('/cluster/nodes', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeNode: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/nodes/${name}/describe`)
    return data
  },

  getComponentStatuses: async (): Promise<any[]> => {
    const { data } = await client.get('/cluster/componentstatuses')
    return data
  },

  getNodePods: async (name: string): Promise<PodInfo[]> => {
    const { data } = await client.get(`/cluster/nodes/${name}/pods`)
    return data
  },

  getNodeEvents: async (name: string): Promise<any[]> => {
    const { data } = await client.get(`/cluster/nodes/${name}/events`)
    return data
  },

  getNodeYaml: async (name: string, forceRefresh: boolean = false): Promise<{ yaml: string }> => {
    const { data } = await client.get(`/cluster/nodes/${name}/yaml`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  applyNodeYaml: async (name: string, yaml: string): Promise<{ status: string }> => {
    const { data } = await client.post(`/cluster/nodes/${name}/yaml/apply`, { yaml })
    return data
  },

  deleteNode: async (name: string): Promise<void> => {
    await client.delete(`/cluster/nodes/${name}`)
  },

  cordonNode: async (name: string): Promise<{ status: string; unschedulable: boolean }> => {
    const { data } = await client.post(`/cluster/nodes/${name}/cordon`)
    return data
  },

  uncordonNode: async (name: string): Promise<{ status: string; unschedulable: boolean }> => {
    const { data } = await client.post(`/cluster/nodes/${name}/uncordon`)
    return data
  },

  drainNode: async (name: string): Promise<{ status: string; drain_id: string }> => {
    const { data } = await client.post(`/cluster/nodes/${name}/drain`)
    return data
  },

  getNodeDrainStatus: async (
    name: string,
    drainId: string,
  ): Promise<{ id: string; node: string; status: string; message?: string | null }> => {
    const { data } = await client.get(`/cluster/nodes/${name}/drain/status`, {
      params: { drain_id: drainId },
    })
    return data
  },

  // Cluster Setup
  getSetupStatus: async (): Promise<{ configured: boolean; mode?: string; secret_name?: string }> => {
    const { data } = await client.get('/auth/setup')
    return data
  },

  submitSetup: async (payload: { mode: 'in_cluster' | 'external'; kubeconfig?: string }) => {
    const { data } = await client.post('/auth/setup', payload)
    return data
  },

  /** 롤아웃 상태 확인 — Setup에서 서비스 재시작 완료 여부 확인 */
  getRolloutStatus: async (): Promise<{ ready: boolean; deployments: Record<string, any> }> => {
    const { data } = await client.get('/auth/setup/rollout-status')
    return data
  },

  // Health check — /health는 /api/v1가 아닌 루트에 있음
  getHealth: async (): Promise<{ status: string; kubernetes: string; openai: string }> => {
    const { data } = await axios.get('/health', {
      baseURL: '',
    })
    return data
  },

  // Advanced Search
  searchResources: async (resourceType: string, namespace?: string, signal?: AbortSignal): Promise<any> => {
    const { data } = await client.get('/cluster/resources', {
      params: {
        resource_type: resourceType,
        all_namespaces: !namespace,
        namespace: namespace || undefined,
        output: 'json',
      },
      timeout: 30000,
      signal,
    })
    return data
  },

  searchMultiResources: async (resourceTypes: string[], namespace?: string): Promise<{ items: any[]; total: number; errors: any[] }> => {
    const { data } = await client.post('/cluster/search', {
      resource_types: resourceTypes,
      namespace: namespace || undefined,
    }, { timeout: 60000 })
    return data
  },

  getApiResources: async (forceRefresh = false): Promise<any[]> => {
    const { data } = await client.get('/cluster/api-resources', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },
}

